# Operasi Matematika & Numpy
Sebelum lanjut, harap:
1. Jika menggunakan jupyter, silahkan pilih "File" --> "Download" --> "Download .ipynb"
2. Jika menggunakan colab, silahkan pilih "File" --> "Save a copy in drive"

In [ ]:
import numpy as np

### Pengenalan array

In [ ]:
#1D array
a = np.array([3, 1, 5, 6])
print('banyaknya entry = ', a.size)
print('dimensi array = ', a.ndim)
print('ukuran array = ', a.shape)

In [ ]:
#Slicing
a[1:]

In [ ]:
print('a[1:3] = ', a[1:3])
print('a[:3] = ', a[:3])

In [ ]:
#2D Array (Matriks)
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(A)

Kodingan di atas sudah benar dan dapat dijalankan. Berikut adalah output yang dihasilkan:

In [ ]:
banyaknya entry =  4
dimensi array =  1
ukuran array =  (4,)
a[1:3] =  [1 5]
a[:3] =  [3 1 5]
[[1 2 3]
 [4 5 6]
 [7 8 9]]

###Latihan-2 (2D array)

In [ ]:
#Lengkapi kodingan berikut
print('banyaknya entry = ', A.size)
print('dimensi array = ', A.ndim)
print('ukuran array = ', A.shape)

In [ ]:
#Slicing
print('baris ke-2 = ', A[1])
print('entry ke-2 & 3 baris ke-2 = ', A[1, 1:])
print('entry pertama & 2 kolom ke-2 = ', A[:2, 1])

###Latihan-3 (Slicing)

In [ ]:
print('baris ke-3 = ', A[2])
print('entry ke-1 & 2 baris ke-3 = ', A[2, 0:2])
print('entry ke-2 & 3 kolom ke-3 = ', A[0:2, 2])

output jawaban:
baris ke-3 =  [7 8 9]

entry ke-1 & 2 baris ke-3 =  [7 8]

entry ke-2 & 3 kolom ke-3 =  [6 9]

### Operasi Matematika di Numpy

In [ ]:
C = A*A
print('perkalian antar entry = ')
print(C)

In [ ]:
print('perkalian matriks = ')
print(np.matmul(A,A))

Kedua operasi yang dijalankan pada kode tersebut adalah perkalian matriks antara A dan A, namun dilakukan dengan menggunakan dua cara yang berbeda, yaitu:

C = A*A: ini adalah cara pertama, di mana kita melakukan perkalian antar entry di antara dua matriks A. Kode ini akan menghasilkan matriks C yang memiliki ukuran yang sama dengan matriks A, di mana setiap elemen pada matriks hasil diperoleh dengan mengalikan elemen yang sesuai pada A dengan elemen yang sesuai pada A.

np.matmul(A,A): ini adalah cara kedua, di mana kita menggunakan fungsi matmul dari NumPy untuk melakukan perkalian matriks antara dua matriks A. Kode ini juga akan menghasilkan matriks yang sama dengan C.

Berikut jawabannya:

In [ ]:
import numpy as np

A = np.array([[1, 2], [7, 8]])

# Cara pertama: perkalian antar entry di A
C = A*A
print('Perkalian antar entry =')
print(C)

# Cara kedua: perkalian matriks antara A dan A
D = np.matmul(A, A)
print('Perkalian matriks =')
print(D)

Perkalian antar entry =
[[ 1  4]
 [49 64]]
Perkalian matriks =
[[15 18]
 [63 78]]


Karena C dan D adalah hasil perkalian matriks yang sama, maka outputnya akan sama untuk kedua cara tersebut.

Bisa dilihat perbedaan perkalian antar entry dan perkalian matriks biasa?

In [ ]:
print('determinan matriks A = ', np.linalg.det(A))
print('determinan matriks C = ', np.linalg.det(C))

In [ ]:
print('nilai dan vektor eigen matriks A = ', np.linalg.eig(A))

In [ ]:
print('nilai eigen matriks A = ', np.linalg.eig(A)[0])

In [ ]:
print('vektor eigen matriks A = ', np.linalg.eig(A)[1])

In [ ]:
e_vec = np.linalg.eig(A)[1][:, 0] #Menagambil eigen vector pertama ([-0.23197, -0.52532209, -0.8186735  ])
e_val = np.linalg.eig(A)[0][0] #Mengambil eigen value pertama (1.61168440e+01)

In [ ]:
print(np.matmul(A, e_vec)) #hasil A * eigen vector

In [ ]:
print(e_val*e_vec) #hasil eigen value * eigen vector (Hasilnya sama)

Ya, karena kita bisa melihat perbedaan antara perkalian antar entry dan perkalian matriks biasa dengan melihat output dari kode yang diberikan.

Pertama, untuk menghitung determinan matriks A dan C, kita menggunakan fungsi det dari NumPy. Namun, perlu diingat bahwa determinan hanya bisa dihitung untuk matriks persegi. Dalam kasus ini, kedua matriks A dan C merupakan matriks persegi, sehingga kita dapat menghitung determinan untuk kedua matriks tersebut.

In [ ]:
import numpy as np

A = np.array([[1, 2], [7, 8]])
C = A*A

# Menghitung determinan
det_A = np.linalg.det(A)
det_C = np.linalg.det(C)

print('Determinan matriks A:', det_A)
print('Determinan matriks C:', det_C)

Determinan matriks A: -6.0
Determinan matriks C: -131.99999999999997


Dalam kasus ini, perbedaan hasil determinan matriks A dan C menunjukkan bahwa operasi perkalian antar entry dan perkalian matriks biasa menghasilkan matriks yang berbeda, sehingga determinan dari kedua matriks tersebut juga berbeda.

Kedua, untuk menghitung nilai eigen dan vektor eigen dari matriks A, kita menggunakan fungsi eig dari NumPy. Fungsi ini akan mengembalikan dua nilai, yaitu nilai eigen dalam bentuk array dan matriks vektor eigen dalam bentuk array 2 dimensi. Kita juga dapat mengambil nilai eigen pertama dan vektor eigen pertama dari matriks A.

In [ ]:
import numpy as np

A = np.array([[1, 2], [7, 8]])

# Menghitung nilai eigen dan vektor eigen
eigenvals, eigenvects = np.linalg.eig(A)

print('Nilai eigen matriks A:', eigenvals)
print('Vektor eigen matriks A:')
print(eigenvects)

# Mengambil eigen value pertama dan eigen vector pertama
e_vec = eigenvects[:, 0]
e_val = eigenvals[0]

# Menghitung A * eigen vector
print(np.matmul(A, e_vec))

# Menghitung eigen value * eigen vector
print(e_val * e_vec)

Nilai eigen matriks A: [-0.62347538  9.62347538]
Vektor eigen matriks A:
[[-0.77640358 -0.22592838]
 [ 0.63023605 -0.97414391]]
[ 0.48406852 -0.39293666]
[ 0.48406852 -0.39293666]


Dalam kasus ini, hasil dari perkalian matriks A dan eigen vector pertama e_vec dan hasil dari perkalian eigen value pertama e_val dan eigen vector pertama e_vec sama, menunjukkan bahwa eigen vector dan eigen value sudah benar-benar merepresentasikan sifat-sifat matriks A.

### Latihan-4 (Operasi matriks)

In [ ]:
A = np.array([[1, 2], [7,8]])
B = np.array([[3, 5], [1,6]])

Hitunglah:
*   2*A
*   2+A
*   A+B
*   A-B
*   determinan A
*   invers dari B, simpan jadi matriks D
*   transpose D
*   nilai eigen dan vektor eigen dari A dan B







In [ ]:
#Ketik Kode disini

import numpy as np

A = np.array([[1, 2], [7, 8]])
B = np.array([[3, 5], [1, 6]])

# 2*A
print("2*A =")
print(2*A)

# 2+A
print("2+A =")
print(2+A)

# A+B
print("A+B =")
print(A+B)

# A-B
print("A-B =")
print(A-B)

# determinan A
print("Determinan A =", np.linalg.det(A))

# invers dari B, simpan jadi matriks D
D = np.linalg.inv(B)
print("Invers dari B, D =")
print(D)

# transpose D
print("Transpose dari D =")
print(D.T)

# nilai eigen dan vektor eigen dari A
eigenvalues, eigenvectors = np.linalg.eig(A)
print("Nilai eigen dan vektor eigen dari A:")
print("Nilai eigen = ", eigenvalues)
print("Vektor eigen = \n", eigenvectors)

# nilai eigen dan vektor eigen dari B
eigenvalues, eigenvectors = np.linalg.eig(B)
print("Nilai eigen dan vektor eigen dari B:")
print("Nilai eigen = ", eigenvalues)
print("Vektor eigen = \n", eigenvectors)


2*A =
[[ 2  4]
 [14 16]]
2+A =
[[ 3  4]
 [ 9 10]]
A+B =
[[ 4  7]
 [ 8 14]]
A-B =
[[-2 -3]
 [ 6  2]]
Determinan A = -6.0
Invers dari B, D =
[[ 0.46153846 -0.38461538]
 [-0.07692308  0.23076923]]
Transpose dari D =
[[ 0.46153846 -0.07692308]
 [-0.38461538  0.23076923]]
Nilai eigen dan vektor eigen dari A:
Nilai eigen =  [-0.62347538  9.62347538]
Vektor eigen = 
 [[-0.77640358 -0.22592838]
 [ 0.63023605 -0.97414391]]
Nilai eigen dan vektor eigen dari B:
Nilai eigen =  [1.8074176 7.1925824]
Vektor eigen = 
 [[-0.97271381 -0.76626443]
 [ 0.23200828 -0.64252535]]


Apa perbedaan dari B*D (perkalian antar entry di B dan D) dan BD (perkalian matriks B dan D)? 

Perkalian antar entry di B dan D dilakukan dengan mengalikan setiap elemen di matriks B dengan elemen yang sesuai di matriks D. Jadi, jika B dan D memiliki ukuran yang sama, hasil dari B*D akan menjadi matriks dengan ukuran yang sama seperti B dan D, di mana setiap elemen pada matriks hasil diperoleh dengan mengalikan elemen yang sesuai pada B dan D.

Contoh:

In [ ]:
import numpy as np

B = np.array([[1, 2], [3, 4]])
D = np.array([[2, 3], [4, 5]])

result = B * D
print(result)

[[ 2  6]
 [12 20]]


Perkalian matriks B dan D (BD)

Perkalian matriks B dan D dilakukan dengan mengalikan setiap baris di matriks B dengan setiap kolom di matriks D. Ukuran dari matriks hasil bergantung pada jumlah kolom dari matriks B dan jumlah baris dari matriks D. Jika jumlah kolom dari B sama dengan jumlah baris dari D, maka operasi perkalian akan valid dan menghasilkan matriks dengan ukuran jumlah baris dari B dan jumlah kolom dari D.

Contoh:

In [ ]:
import numpy as np

B = np.array([[1, 2], [3, 4]])
D = np.array([[2, 3], [4, 5]])

result = B.dot(D)
print(result)

[[10 13]
 [22 29]]


Jadi, dapat dilihat bahwa perbedaan utama antara B*D dan BD adalah bahwa B*D merupakan perkalian antar entry di B dan D, sedangkan BD adalah perkalian matriks antara B dan D.

output jawaban: 

perkalian entry = [[ 1.38461538, -1.92307692], [-0.07692308,  1.38461538]]

perkalian matriks = [[ 1.00000000e+00, -1.11022302e-16],
       [ 1.11022302e-16,  1.00000000e+00]]
